iwyu suggests forward declaration when using typeid

Thanks for all your work on iwyu. I can't imagine sorting out my dependency messes without it :)

I've recently seen an issue where iwyu will suggest a forward-declaration instead of a `#include` for `Dependency` here:

```c++
// user/user.cpp
#include <dependency.h>
#include <typeindex>

auto useTypeIndexOfDependency(const Dependency &dependency) -> void {
  std::type_index type = typeid(dependency);
}
```

```
$ iwyu_tool -p build user/user.cpp

/home/matthew/Documents/iwyu/minimal-example/user/user.cpp should add these lines:
class Dependency;

/home/matthew/Documents/iwyu/minimal-example/user/user.cpp should remove these lines:
- #include <dependency.h>  // lines 1-1

The full include-list for /home/matthew/Documents/iwyu/minimal-example/user/user.cpp:
#include <typeindex>  // for type_index
class Dependency;
---
```
If I follow its suggestion, I get 
```
[ 50%] Building CXX object user/CMakeFiles/user.dir/user.cpp.o
/home/matthew/Documents/iwyu/minimal-example/user/user.cpp: In function ‘void useTypeIndexOfDependency(const Dependency&)’:
/home/matthew/Documents/iwyu/minimal-example/user/user.cpp:5:33: error: invalid use of incomplete type ‘class Dependency’
    5 |   std::type_index type = typeid(dependency);
      |                                 ^~~~~~~~~~
/home/matthew/Documents/iwyu/minimal-example/user/user.cpp:2:7: note: forward declaration of ‘class Dependency’
    2 | class Dependency;
      |       ^~~~~~~~~~
make[2]: *** [user/CMakeFiles/user.dir/build.make:82: user/CMakeFiles/user.dir/user.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:134: user/CMakeFiles/user.dir/all] Error 2
make: *** [Makefile:103: all] Error 2
```
I believe this is because you cannot apply `typeid` to an incomplete type.

I tested this with
```
$ include-what-you-use --version
include-what-you-use 0.15 based on Debian clang version 11.0.1-2
```
which is the current version on Debian Stable.

I've uploaded a full minimal example [here](https://github.com/cottonxray/iwyu-typeid-potential-issue).

I would be happy to spend some time to contribute a fix for this, if that would help. Equally, if there's any extra information that I can provide, then let me know.

But I wanted to check:
* do you also think it's an issue, or have I missed something?
* if you'd like me to work on a fix - where would be the best place to make this change in the code?

Thanks again for all your work on this tool.

--- a/iwyu.cc
+++ b/iwyu.cc
@@ -2549,6 +2549,23 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
     return true;
   }
 
+  // The C++ spec does not allow us to apply typeid to an incomplete type.
+  // Therefore, we must report the type to be included.
+  bool VisitCXXTypeidExpr(clang::CXXTypeidExpr* expr) {
+    if (CanIgnoreCurrentASTNode())
+      return true;
+
+    QualType type;
+    if (expr->isTypeOperand()) {
+      type = expr->getTypeOperandSourceInfo()->getType();
+    } else {
+      type = expr->getExprOperand()->getType();
+    }
+
+    ReportTypeUse(CurrentLoc(), type.getTypePtr());
+    return true;
+  }
+
   //------------------------------------------------------------
   // Visitors of types derived from clang::Type.
 

--- a/tests/cxx/typeid.cc
+++ b/tests/cxx/typeid.cc
@@ -1,0 +1,48 @@
+//===--- typeid.cc - test input file for iwyu -----------------------------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+
+// IWYU_ARGS: -I .
+
+// Tests that IWYU suggests an include instead of a forward declaration for
+// expressions used as arguments to typeid(). This is necessary as calling
+// typeid() on an incomplete type causes the program to be ill-formed.
+// We need both typeid(expr) and typeid(type) calls to cause the necessary files
+// to be included, rather than forward declared.
+
+#include "tests/cxx/direct.h"
+#include <typeinfo>
+
+// Tests that typeid(expr) triggers an include, rather than a
+// forward declaration.
+// IWYU: IndirectClass needs a declaration
+void GetTypeInfoOfIndirectExpr(const IndirectClass &indirect) {
+  // IWYU: IndirectClass is...*indirect.h
+  const std::type_info &typeInfo = typeid(indirect);
+}
+
+// Tests that typeid(type) triggers an include, rather than a
+// forward declaration.
+void GetTypeInfoOfIndirectType() {
+  // IWYU: IndirectClass is...*indirect.h
+  const std::type_info &typeInfo = typeid(IndirectClass);
+}
+
+/**** IWYU_SUMMARY
+
+tests/cxx/typeid.cc should add these lines:
+#include "tests/cxx/indirect.h"
+
+tests/cxx/typeid.cc should remove these lines:
+- #include "tests/cxx/direct.h"  // lines XX-XX
+
+The full include-list for tests/cxx/typeid.cc:
+#include <typeinfo>  // for type_info
+#include "tests/cxx/indirect.h"  // for IndirectClass
+
+***** IWYU_SUMMARY */

